### ISSUE TITLE
[SPIR-V] Avoid OpDemoteToHelperInvocation for Vulkan 1.3 target

### ISSUE THREAD

**Issue by webanck** [microsoft/DirectXShaderCompiler#5635]
Hello, it seems there has been some regression on the `OpDemoteToHelperInvocation` matter.

Take the following pixel shader example:
```hlsl
float4 pixelShader(float4 p : SV_POSITION) : SV_TARGET
{
	if (p.x == 0.) discard;
    return float4(0., 0., 0., 0.);
}
```

Compiling it with DXC version `1.7 - 1.7.0.4004 (4cfac0142)`: `.\dxc.exe -E pixelShader -T ps_6_2  -spirv -fspv-target-env="vulkan1.3" .\test.hlsl` outputs the following:
```
; SPIR-V
; Version: 1.6
; Generator: Google spiregg; 0
; Bound: 19
; Schema: 0
               OpCapability Shader
               OpCapability DemoteToHelperInvocation
               OpMemoryModel Logical GLSL450
               OpEntryPoint Fragment %pixelShader "pixelShader" %gl_FragCoord %out_var_SV_TARGET
               OpExecutionMode %pixelShader OriginUpperLeft
               OpSource HLSL 620
               OpName %out_var_SV_TARGET "out.var.SV_TARGET"
               OpName %pixelShader "pixelShader"
               OpDecorate %gl_FragCoord BuiltIn FragCoord
               OpDecorate %out_var_SV_TARGET Location 0
      %float = OpTypeFloat 32
    %float_0 = OpConstant %float 0
    %v4float = OpTypeVector %float 4
          %7 = OpConstantComposite %v4float %float_0 %float_0 %float_0 %float_0
%_ptr_Input_v4float = OpTypePointer Input %v4float
%_ptr_Output_v4float = OpTypePointer Output %v4float
       %void = OpTypeVoid
         %11 = OpTypeFunction %void
       %bool = OpTypeBool
%gl_FragCoord = OpVariable %_ptr_Input_v4float Input
%out_var_SV_TARGET = OpVariable %_ptr_Output_v4float Output
%pixelShader = OpFunction %void None %11
         %13 = OpLabel
         %14 = OpLoad %v4float %gl_FragCoord
         %15 = OpCompositeExtract %float %14 0
         %16 = OpFOrdEqual %bool %15 %float_0
               OpSelectionMerge %17 None
               OpBranchConditional %16 %18 %17
         %18 = OpLabel
               OpDemoteToHelperInvocation
               OpBranch %17
         %17 = OpLabel
               OpStore %out_var_SV_TARGET %7
               OpReturn
               OpFunctionEnd
```

As you can see, it outputs an `OpDemoteToHelperInvocation` operation near the end. 

This doesn't happen with Vulkan target versions 1.0, 1.1, and 1.2 (`OpKill` is used instead of `OpDemoteToHelperInvocation`); it starts at target 1.3. So, is there any way to prevent those `OpDemoteToHelperInvocation` operations to be output for the Vulkan target 1.3?

_Originally posted by @webanck in https://github.com/microsoft/DirectXShaderCompiler/issues/2823#issuecomment-1701265093_

**Comment by sudonatalie** [comment]
`VK_EXT_shader_demote_to_helper_invocation` has been promoted to [Vulkan 1.3](https://registry.khronos.org/vulkan/specs/1.3-extensions/html/vkspec.html#versions-1.3-promotions) core API, so we don't require the extension to be explicitly enabled in order to use the `OpDemoteToHelperInvocation` instruction when the target is 1.3. As far as I can see this is the correct behavior. Can you provide any additional context to explain why you are trying to avoid this instruction?

**Comment by webanck** [comment]
Thanks for your answer @sudonatalie.

Indeed, am aware that the `VK_EXT_shader_demote_to_helper_invocation` extension is part of the Vulkan 1.3 core API. I was however mislead by the existence of the `VkPhysicalDeviceShaderDemoteToHelperInvocationFeatures` which allows checking for the support of the demote instruction. I previously had some troubles with extensions part of the core API, but nonetheless unsupported by a given GU/driver combo.

Now, re-reading carefully [some of] the documentation, extensions features that are part of the core API are sometimes optional, but it is then stated in the documentation. An example is the `VK_KHR_shader_float16_int8` extension, core since Vulkan 1.2, with optional support of shaderFloat16 and shaderInt8 capabilities. But in the case at hand, in the documentation of `VK_EXT_shader_demote_to_helper_invocation`, there is no mention of anything optional.

TLDR: if I understand the documentation correctly, you are right and the support Vulkan 1.3 encompasses the support of the SPIRV demote op.
